### PR TITLE
Fix deduplication on address changes.

### PR DESCRIPTION
**Pull request by boaks**
Fix issue #1082 .

Signed-off-by: Achim Kraus <achim.kraus@bosch-si.com>

### CODE REVIEW

File: californium-core/src/main/java/org/eclipse/californium/core/network/UdpMatcher.java [hunk 1]
```diff
@@ -219,9 +219,8 @@ public void receiveRequest(final Request request, final EndpointReceiver receive
 			duplicate = endpointContextMatcher.isToBeSent(previousSourceContext, sourceContext);
 			if (!duplicate) {
 				// the new context doesn't match the previous.
-				exchangeStore.remove(idByMID, previous);
-				if (exchangeStore.findPrevious(idByMID, exchange) != null) {
-					LOGGER.warn("new request could not be registered!");
+				if (!exchangeStore.replacePrevious(idByMID, previous, exchange)) {
+					LOGGER.warn("new request {} could not be registered!", request);
```
Comment (sbernard31) [inline californium-core/src/main/java/org/eclipse/californium/core/network/UdpMatcher.java]: could we add a log about the successful replacement  with old and new request ?

File: californium-core/src/main/java/org/eclipse/californium/core/network/deduplication/CropRotation.java [hunk 2]
```diff
@@ -114,14 +113,20 @@ public Exchange findPrevious(KeyMID key, Exchange exchange) {
 		return prev;
 	}
 
+	@Override
+	public boolean replacePrevious(KeyMID key, Exchange previous, Exchange exchange) {
+		int s = second;
+		return maps[s].replace(key, previous, exchange) || maps[s].putIfAbsent(key, exchange) == null;
```
Comment (sbernard31) [inline californium-core/src/main/java/org/eclipse/californium/core/network/deduplication/CropRotation.java]: Maybe we don't care about CropRotation but I don't get why in findPrevious we write in f then in s and here we only write in s.

File: californium-core/src/main/java/org/eclipse/californium/core/network/UdpMatcher.java (hunk 1 above)
Comment (boaks) [reply californium-core/src/main/java/org/eclipse/californium/core/network/UdpMatcher.java]: OK, I'll add it.

**Comment by sbernard31** [comment]
LGTM.

**Comment by boaks** [comment]
Logging message added.

**Pull request merged**